src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -1049,113 +1045,156 @@ public bool TakeStartLine(ReadableBuffer buffer, out ReadCursor consumed, out Re
-                {
-                    RejectRequestLine(start, end);
-                }
-            }
-
+            var pathStart = -1;
+            var queryStart = -1;
+            var queryEnd = -1;
+            var pathEnd = -1;
+            var versionStart = -1;
             var queryString = "";
-            var queryEnd = pathEnd;
-            if (questionMarkIndex != -1)
-            {
-                queryEnd = spaceIndex;
-                if (spaceIndex == -1)
-                {
-                    RejectRequestLine(start, end);
-                }
+            var httpVersion = "";
+            var method = "";
+            var state = StartLineState.ExpectKnownMethod;
 
-                queryString = pathToEndSpan.Slice(pathEnd, queryEnd - pathEnd).GetAsciiString();
-            }
-
-            if (pathBegin == pathEnd)
-            {
-                RejectRequestLine(start, end);
-            }
-
-            var versionBegin = queryEnd + 1;
-            var versionToEndSpan = pathToEndSpan.Slice(versionBegin, pathToEndSpan.Length - versionBegin);
-            versionBegin = 0;
-            var versionEnd = versionToEndSpan.IndexOf(ByteCR);
-
-            if (versionEnd == -1)
+            fixed (byte* data = &span.DangerousGetPinnableReference())
             {
-                RejectRequestLine(start, end);
-            }
-
-            if (!versionToEndSpan.GetKnownVersion(out string httpVersion))
-            {
-                httpVersion = versionToEndSpan.Slice(0, versionEnd).GetAsciiStringEscaped();
-
-                if (httpVersion == string.Empty)
-                {
-                    RejectRequestLine(start, end);
-                }
-                else
+                var length = span.Length;
+                for (var i = 0; i < length; i++)
                 {
-                    RejectRequest(RequestRejectionReason.UnrecognizedHTTPVersion, httpVersion);
+                    var ch = data[i];
+
+                    switch (state)
+                    {
+                        case StartLineState.ExpectKnownMethod:
+                            if (span.GetKnownMethod(out method))
+                            {
+                                // Update the index, current char, state and jump directly
+                                // to the next state
+                                i += method.Length + 1;
+                                ch = data[i];
+                                state = StartLineState.ExpectPath;
+
+                                goto case StartLineState.ExpectPath;
+                            }
+
+                            state = StartLineState.ExpectUnknownMethod;
+                            goto case StartLineState.ExpectUnknownMethod;
+
+                        case StartLineState.ExpectUnknownMethod:
+                            if (ch == ByteSpace)
+                            {
+                                method = span.Slice(0, i).GetAsciiString();
```

Comment:
Also, it's ugly because we have a fast path that does materialization inline so I'm not sure it makes sense here. There would need to be a check for if method == null.